swarm/storage/pusher/storer.go

```diff
@@ -0,0 +1,90 @@
+package pusher
+
+import (
+	"context"
+
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/swarm/chunk"
+	"github.com/ethereum/go-ethereum/swarm/log"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+	"github.com/ethereum/go-ethereum/swarm/storage/localstore"
+)
+
+// Store is the storage interface to save chunks
+// NetStore implements this interface
+type Store interface {
+	Put(context.Context, chunk.ModePut, chunk.Chunk) error
+}
+
+// storer is the object used by the push-sync server side protocol
+type storer struct {
+	store Store  // store to put chunks in, and retrieve them
+	ps    PubSub // pubsub interface to receive chunks and send receipts
+}
+
+// RunStorer starts a protocol run by storer nodes to handle push-synced chunks
+// that fall within their area of responsibility.
+// The protocol makes sure that
+// - the chunks are stored and synced to their nearest neighbours and
+// - a statement of custody receipt is sent as a response to the originator
+// it returns a cancel function that deregisters the handler
+func RunStorer(store *localstore.DB, ps PubSub) func() {
+	s := &storer{
+		store: store,
+		//store.NewPutter(localstore.ModePutSync),
+		ps: ps,
+	}
+	// registers a handler on pssChunkTopic
+	deregister := s.ps.Register(pssChunkTopic, func(msg []byte, _ *p2p.Peer) error {
+		return s.handleChunkMsg(msg)
+	})
+	return deregister
+}
+
+// handleChunkMsg is called by the pss dispatcher on pssChunkTopic msgs
+// - deserialises chunkMsg and
+// - calls storer.processChunkMsg function
+func (s *storer) handleChunkMsg(msg []byte) error {
+	chmsg, err := decodeChunkMsg(msg)
+	if err != nil {
+		return err
+	}
+	log.Error("Handler", "chunk", label(chmsg.Addr), "origin", label(chmsg.Origin), "self", s.ps.BaseAddr())
+	return s.processChunkMsg(chmsg)
+}
+
+// processChunkMsg processes a chunk received vie pss pssChunkTopic
+// these chunk messages are sent to their address as destination
+// using neighbourhood addressing. Therefore nodes only handle
+// chunks that fall within their area of responsibility.
+// Upon receiving the chunk is saved and a statement of custody
+// receipt message is sent as a response to the originator.
+func (s *storer) processChunkMsg(chmsg *chunkMsg) error {
+	// TODO: double check if it falls in area of responsibility
+	ch := storage.NewChunk(chmsg.Addr, chmsg.Data)
+	err := s.store.Put(context.TODO(), chunk.ModePutSync, ch)
+	if err != nil {
+		return err
+	}
+	// TODO: check if originator or relayer is a nearest neighbour then return
+	// otherwise send back receipt
```

Comment:
if originator is indeed nearest neighbor how does it know it should not keep sending?